tempesta_fw/http_sess.c

```diff
@@ -0,0 +1,731 @@
+	tfw_str_dprint(ua, "\t...User-Agent");				\
+} while (0)
+#else
+#define TFW_PRINT_STICKY_COOKIE(addr, ua, sv)
+#endif
+
+/*
+ * Create Tempesta sticky cookie value.
+ *
+ * Tempesta sticky cookie is based on:
+ * - HTTP request source IP address;
+ * - HTTP request User-Agent string;
+ * - Current timestamp;
+ * - The secret key;
+ */
+static int
+__sticky_calc(TfwHttpReq *req, StickyVal *sv)
+{
+	int r, addr_len;
+	TfwStr ua_value = { 0 };
+	TfwAddr *addr = &req->conn->peer->addr;
+	TfwStr *hdr, *c, *end;
+	char desc[sizeof(struct shash_desc)
+		  + crypto_shash_descsize(tfw_sticky_shash)]
+		  CRYPTO_MINALIGN_ATTR;
+	struct shash_desc *shash_desc = (struct shash_desc *)desc;
+
+	/* User-Agent header field is not mandatory and may be missing. */
+	hdr = &req->h_tbl->tbl[TFW_HTTP_HDR_USER_AGENT];
+	if (!TFW_STR_EMPTY(hdr))
+		tfw_http_msg_clnthdr_val(hdr, TFW_HTTP_HDR_USER_AGENT,
+					 &ua_value);
+
+	addr_len = tfw_addr_sa_len(addr);
+
+	memset(desc, 0, sizeof(desc));
+	shash_desc->tfm = tfw_sticky_shash;
+	shash_desc->flags = 0;
+
+	TFW_PRINT_STICKY_COOKIE(addr, &ua_value, sv);
+
+	if ((r = crypto_shash_init(shash_desc)))
+		return r;
+	if ((r = crypto_shash_update(shash_desc, (u8 *)&addr->sa, addr_len)))
+		return r;
+	if (ua_value.len) {
+		TFW_STR_FOR_EACH_CHUNK(c, &ua_value, end) {
+			r = crypto_shash_update(shash_desc, c->ptr, c->len);
+			if (r)
+				return r;
+		}
+	}
+	return crypto_shash_finup(shash_desc, (u8 *)&sv->ts, sizeof(sv->ts),
+				  sv->hmac);
+}
+
+static int
+tfw_http_sticky_calc(TfwHttpReq *req, StickyVal *sv)
+{
+	sv->ts = jiffies;
+
+	return __sticky_calc(req, sv);
+}
+
+/*
+ * Add Tempesta sticky cookie to an HTTP response.
+ *
+ * Create a complete 'Set-Cookie:' header field, and add it
+ * to the HTTP response' header block.
+ */
+static int
+tfw_http_sticky_add(TfwHttpResp *resp, TfwHttpReq *req)
+{
+	static const unsigned int len = sizeof(StickyVal) * 2;
+	unsigned int r;
+	TfwHttpSess *sess = req->sess;
+	unsigned long ts_be64 = cpu_to_be64(sess->ts);
+	char buf[len];
+	TfwStr set_cookie = {
+		.ptr = (TfwStr []) {
+			{ .ptr = S_F_SET_COOKIE, .len = SLEN(S_F_SET_COOKIE) },
+			{ .ptr = tfw_cfg_sticky.name_eq.ptr,
+			  .len = tfw_cfg_sticky.name_eq.len },
+			{ .ptr = buf, .len = len },
+		},
+		.len = SLEN(S_F_SET_COOKIE) + tfw_cfg_sticky.name_eq.len + len,
+		.eolen = 2,
+		.flags = 3 << TFW_STR_CN_SHIFT
+	};
+
+	/* See comment from tfw_http_sticky_send_302(). */
+	bin2hex(buf, &ts_be64, sizeof(ts_be64));
+	bin2hex(&buf[sizeof(ts_be64) * 2], sess->hmac, sizeof(sess->hmac));
+
+	TFW_DBG("%s: \"" S_F_SET_COOKIE "%.*s=%.*s\"\n", __func__,
+		PR_TFW_STR(&tfw_cfg_sticky.name), len, buf);
+
+	r = tfw_http_msg_hdr_add((TfwHttpMsg *)resp, &set_cookie);
+	if (r)
+		TFW_WARN("Cannot add \"" S_F_SET_COOKIE "%.*s=%.*s\"\n",
+			 PR_TFW_STR(&tfw_cfg_sticky.name), len, buf);
+	return r;
+}
+
+/*
+ * No Tempesta sticky cookie found.
+ *
+ * Create Tempesta sticky cookie value, and store it for future use.
```

Comment:
This comment is no longer valid for this function. That's done in `tfw_http_sess_obtain()` now.